llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -199,93 +170,151 @@ groupEntryPointsByKernelType(ModuleDesc &MD,
   return EntryPointGroups;
 }
 
-// Represents a call graph between functions in a module. Nodes are functions,
-// edges are "calls" relation.
-class CallGraph {
+// Represents "dependency" or "use" graph of global objects (functions and
+// global variables) in a module. It is used during device code split to
+// understand which global variables and functions (other than entry points)
+// should be included into a split module.
+//
+// Nodes of the graph represent LLVM's GlobalObjects, edges "A" -> "B" represent
+// the fact that if "A" is include into a module, then "B" should be included
+// as well.
+//
+// Examples of dependencies which are represented in this graph:
+// - Function FA calls function FB
+// - Function FA uses global variable GA
+// - Global variable GA references (intiialized with) function FB
+// - Function FA stores address of a function FB somewhere
+// - Function FA uses global variable GB
+//
+// The following cases are treated as dependencies between global objects:
+// 1. Global object A is used within by a global object B in any way (store,
+//    bitcast, phi node, call, etc.): "A" -> "B" edge will be added to the
+//    graph;
+// 2. function A performs an indirect call of a function with signature S and
```

Comment:
If the caller function had the same signature S, will it also be included here?

Thanks

Reply:
Yes. The proposed version contains no filter of potential callees (except them being kernels) and therefore would include the caller itself as well, if it has signature matching indirect call.

This shouldn't cause any issues, because the caller will anyway be included if we reached it through call graph; and it won't be included otherwise (i.e. if we look at a graph, which includes some potentially indirectly callable callee, but not indirect calls)